autora/theorist/bsr/funcs.py

```diff
@@ -0,0 +1,927 @@
+import copy
+from enum import Enum
+from functools import wraps
+from typing import Callable, Dict, List, Optional, Tuple, Union, cast
+
+import numpy as np
+import pandas as pd
+from scipy.stats import invgamma, norm
+
+from .node import Node, NodeType
+
+
+def check_empty(func: Callable):
+    """
+    A decorator that, if applied to `func`, checks whether an argument in `func` is an
+    un-initialized node (i.e. node.node_type == NodeType.Empty). If so, an error is raised.
+    """
+
+    @wraps(func)
+    def func_wrapper(*args, **kwargs):
+        for arg in args:
+            if isinstance(arg, Node):
+                if arg.node_type == NodeType.EMPTY:
+                    raise TypeError(
+                        "uninitialized node found in {}".format(func.__name__)
```

Comment:
Users of AutoRA do not access the `node` data structure itself in most cases as all they concern about should be the expression tree and the expression it represents. They will also rarely need to create a node and initialize it. In the extreme case as you said so, nothing malicious or impactful would occur. There might be some error raised for null pointer since we might do something in the code like
```
if node.node_type == NodeType.LEAF:
     bar()
elif node.node_type == NodeType.UNARY:
     foo()
else:
     baz()
```
and the `baz()` would assume the node to be binary if we have this decorator applied.